### PR TITLE
Using LoadLibraryEx and LOAD_LIBRARY_SEARCH_* flag for loading DLLs o…

### LINKED ISSUE #37209: Pytorch upgrade 1.4 -> 1.5 broken on Windows
## Issue description

Just upgraded to 1.5 on Windows 10 using conda, and pytorch fails to import with the following error:
Python 3.6.9 |Anaconda, Inc.| (default, Jul 30 2019, 14:00:49) [MSC v.1915 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "(user dir)\Anaconda3\envs\cenv\lib\site-packages\torch\__init__.py", line 81, in <module>
    ctypes.CDLL(dll)
  File "(user dir)\Anaconda3\envs\cenv\lib\ctypes\__init__.py", line 348, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: [WinError 126] The specified module could not be found

## Code example

import torch

## System Info

collect_env.py fails due to import torch failure above.

- Installed PyTorch via conda, from 1.4 with cudatoolkit=10.1, to 1.5 with cudatoolkit=10.2, with the following command:
conda install pytorch torchvision cudatoolkit=10.2 -c pytorch
- Windows 10
- Python version: 3.6.9
- CUDA/cuDNN version: 10.2

cc @peterjc123 @nbcsm @guyang3532

### PR DESCRIPTION
**Pull request by peterjc123**
…n Windows

Without this PR, the OS try to find the DLL in the following directories.
- The directory from which the application loaded.
- The system directory. Use the GetSystemDirectory function to get the path of this directory.
- The 16-bit system directory. There is no function that obtains the path of this directory, but it is searched.
- The Windows directory. Use the GetWindowsDirectory function to get the path of this directory.
- The current directory.
- The directories that are listed in the PATH environment variable. Note that this does not include the per-application path specified by the App Paths registry key. The App Paths key is not used when computing the DLL search path.

If we use  LoadLibraryEx with LOAD_LIBRARY_SEARCH_* flags, the directories are searched in the following order.

- The directory that contains the DLL (LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR). This directory is searched only for dependencies of the DLL to be loaded.
- The application directory (LOAD_LIBRARY_SEARCH_APPLICATION_DIR).
- Paths explicitly added to the application search path with the AddDllDirectory function (LOAD_LIBRARY_SEARCH_USER_DIRS) or the SetDllDirectory function. If more than one path has been added, the order in which the paths are searched is unspecified.
- The System32 directory (LOAD_LIBRARY_SEARCH_SYSTEM32).

Advantages:
1. The directory that contains the DLL comes first and it's desirable for us, because the dependencies in `lib` should always be preferred.
2. The system directory is considered in the last place. According to some of the bug reports, the DLL load failure are caused by loading the conflicting ones in systemroot.

Neural:
1. The directories in `PATH` are not considered. Similar things happen as described in the previous point. So it may be beneficial for normal users. However, it may cause failures if there are some new dependencies if built from source. (Resolved by making the fallback to `LoadLibraryW` if error code is `126`)

Disadvantages:
1. LoadLibraryEx with LOAD_LIBRARY_SEARCH_* flags is only available for Win7/2008 R2 + KB2533623 and up. (Resolved by making the fallback to `LoadLibraryW` if it is not supported)
2. Failure during the call of `LoadLibraryEx` will lead to the OS to pop up a modal dialog, which can block the process if user is using a CLI-only interface. This can be switched off by calling `SetErrorMode`. (Resolved by calling `SetErrorMode`)

Test plan:
Test some common cases (in a new repo maybe) including
1. Python 3.6/3.7/3.8, conda python, conda install
2. Python 3.6/3.7/3.8, conda python, pip install
3. Python 3.6/3.7/3.8, official python, pip install
Plus some corner cases like
1. Conflicting DLLs in systemroot or `PATH`
2. Remove some local dependencies and use global ones

References:
1. https://docs.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-seterrormode
2. https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibraryexa
3. https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order#standard-search-order-for-desktop-applications

What do you think, @malfet @ezyang ?

### CODE REVIEW

**Comment by ezyang** [comment]
Maybe @malfet is more opinionated here, but the proposed new behavior seems reasonable: I'm willing to give it a try. I hate that there's not really a "standard" in Windows for this sort of thing; it seems like everyone does something different :>

File: torch/__init__.py [hunk 1]
```diff
@@ -62,22 +62,27 @@
...
+    for dll_path in dll_paths:
+        if sys.version_info >= (3, 8):
             os.add_dll_directory(dll_path)
-
-    if is_conda or sys.version_info < (3, 8):
-        dll_paths = [th_dll_path, py_dll_path, nvtoolsext_dll_path, cuda_path]
-        dll_paths = list(filter(os.path.exists, dll_paths)) + [os.environ['PATH']]
-
-        os.environ['PATH'] = ';'.join(dll_paths)
+        else:
```
Comment (malfet) [inline torch/__init__.py:71]: To be super careful here and support older versions of windows, why not check that kernel32 actually has the method
```suggestion
        elif hasattr(kernel32, 'AddDllDirectory'):
```

File: torch/__init__.py [hunk 2]
```diff
@@ -62,22 +62,27 @@
...
+            if res == 0:
+                err = ctypes.WinError(ctypes.get_last_error())
+                err.strerror += ' Error adding "{}" to the DLL directories.'.format(dll_path)
+                raise err
 
     import glob
     dlls = glob.glob(os.path.join(th_dll_path, '*.dll'))
     for dll in dlls:
-        ctypes.CDLL(dll)
+        res = kernel32.LoadLibraryExW(dll, 0, 0x00001100)
```
Comment (malfet) [inline torch/__init__.py:81]: Same as above here:
```
    if hasattr('kernel32', LoadLibraryExW'):
        res = kernel32.LoadLibraryExW(dll, 0, 0x1100)
    else:
        ctypes.CDLL(dll)
````

Comment (peterjc123) [reply torch/__init__.py:81]: well, the function itself is supported even in xp but the arguments isn't. I'll try to look for ways to check that.

File: torch/__init__.py (hunk 1 above)
Comment (peterjc123) [reply torch/__init__.py:71]: This one sounds reasonable.

**Comment by peterjc123** [comment]
> Maybe @malfet is more opinionated here, but the proposed new behavior seems reasonable: I'm willing to give it a try. I hate that there's not really a "standard" in Windows for this sort of thing; it seems like everyone does something different :>

Well, I guess it is because none of the current plans could be considered perfect.

**Comment by malfet** [comment]
Looks good to me!

**Comment by peterjc123** [comment]
Verified by the tests in https://github.com/peterjc123/pytorch_dll_load_smoketests. Marking it as ready.

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
This PR were reverted, as it caused TestTypeHints failure, see for example https://app.circleci.com/pipelines/github/pytorch/pytorch/167977/workflows/2cf6c758-57a3-4a45-a465-01dfd7832a31/jobs/5432835/tests

**Comment by peterjc123** [comment]
@malfet Could you please tell me how to deal with it?

**Pull request closed without merging**